_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+---
+layout: post
+title: A story unfolds
+category: Haskell
+description: or how I tried to plant a tree
+---
+# A story unfolds
+
+### or how I tried to plant a tree
+
+I tried to re-implement the Unix `tree` tool in Haskell, as a learning exercise. The goal is to create a tool that, when provided with a directory, lists all the contents and prints them in a way similar to this:
+```
+$ tree
+.
+├── a
+│   └── c
+└── b
+    ├── d
+    └── e
+
+5 directories, 0 files
+```
+It is well-defined, well-structured, and obviously recursive problem, so it should be easy to do, a no-brainer almost[^excuses].
+
+I started hacking away at the tree.
+
+The thing I would need first was a function to list directory contents, and another one to let me distinguish between directories and files. Quite fortunately, I had a pair of trusty axes on hand[^axes]:
+```haskell
+listDir :: File -> IO [File]
+isDir :: File -> IO Boolean
+```
+A naïve implementation came to mind right away[^madeup]:
+```haskell    
+tree :: File -> IO ()
+tree file = tree' "" file
+  where tree' prefix file = do
+          putStrLn (prefix ++ "- " ++ name file)
+          isDirectory <- isDir file
+          if isDirectory
+          then do
+            contents <- listDir file
+            traverse_ (tree' (' ':prefix)) contents
+          else return ()
+```
+It does not produce nice branch-looking lines between filenames, relying on indentation instead - but it is close enough.
+
+Obviously, this code is clumsy - and not just because the utility functions are crude. No, this code is just clumsy and inflexible in itself - and what's even worse, it makes one cardinal sin.
+_It mixes the responsibilities._
+
+While it is not the gravest sin of them all, as one could argue that _traversing_ the filesystem and _printing_ the filenames are very strongly related, it would be nice to do these two separately. In fact, the added flexibility might come in handy if we want to do something more than just printing - and, as we're about to see, at pretty much no added cost.
+
+So into the forest we go, preparing our tree-like data structure:
+```haskell
+data FileTree = RegularFile File | Directory File [FileTree]
```

Comment:
Does `RegularFile File` mean _a RegularFile, which IS a File_ or _a RegularFile, which HAS a File`? and even more importantly, does _Directory File [FileTree]_ mean _a Directory which IS?HAS a File and HAS? a list of FileTrees_?

Reply:
Judging from the uses, it seems it's all _HAS_

Reply:
Is there any inheritance-like concept in Haskell, other than type classes (kinda)?

Reply:
Well, turns out my naming is unfortunate. A `RegularFile` is a node in a tree, and it _is_ a file (or rather represents a file) that _has_ its own path, while `Directory` _is_ a file that _has_ a path and contains some other stuff (file trees, since the contents might be directories as well). I should've named `File` differently, like `FileHandle`, `FileName` or just `Path` - will fix that.

There is no inheritance in Haskell, there even is no subtyping - however one could argue that the type constraints are the closest thing to subtyping. Type constraints are like context bounds in Scala, basically requiring that a certain type has a typeclass instance defined - see the `treeBuilderM` type signature:
```haskell
treeBuilderM :: Monad m => LookAroundM m a -> BuildNodeM t m a -> a -> m (t a)
```
The `Monad m` part requires that the `m` type (or _type constructor_, rather) is a monad. But then a constraint can be applied to a typeclass definition:
```haskell
class Functor a => Applicative a where
  --...
```
which states that `a` has to be a `Functor` in order to be an `Applicative`. Since applicatives are a strict subset of functors, we could say that there is a subtype-ish relationship in place - but it's a relationship between _typeclasses_ and not _types_.

Reply:
Maybe `File` could be itself renamed to `Node`? but actually, the issue stems from unfortunate naming back in Unix itself, as "file" can mean both a regular file, and node of filesystem, even a directory (named pipes, symlinks etc.)... and this is implied very indirectly.

https://docs.oracle.com/javase/7/docs/api/java/io/File.html#isFile() :trollface: